### DISCUSSION TITLE
Explain SqsListener concurrency

### DISCUSSION THREAD

**Discussion by andrej-urvantsev** [awspring/spring-cloud-aws#1203] [Q&A] [answered]
My use case is quite simple - read messages from SQS, transform them and write into relational database(jdbc).

Problem is that with all defaults, SqsListener starts 10 threads(I guess it's defined by maxConcurrentMessages) and therefore size of batches being read from SQS is a quite small - it's very rare when it's more than 1 message.

But 10 threads writing into DB one record take 10 DB connections. And writing just one record per connection is less than optimal.

So, I would prefer less concurrent threads for SqsListener, but bigger batches(even if it takes more polls). And here I struggle to understand how things work.

I would assume that logical thing to do is to set `maxConcurrentMessages=2` and `messagesPerPoll=30`, but that is not possible, application fails with `messagesPerPoll should be less than or equal to maxConcurrentMessages`, because `maxConcurrentMessages` meaning is actually "max inflight messages".

So, how do I maximize batch size and minimize concurrency?

**Comment by maciejwalkowiak** [comment]
@lazystone it looks like batching consumer would fit your use case well. Have you had a look at https://docs.awspring.io/spring-cloud-aws/docs/3.2.0/reference/html/index.html#batch-processing ?

**Reply by andrej-urvantsev** [reply]
My understanding is that if I use something like this:

```java
    @SqsListener(queueNames = {"${application.incoming-events-queue}"})
    public void handle(List<Message<String>> messages) {
       //
    }
```
then it should be in BATCH listening mode? Or should I in addition configure it explicitly in container?

**Reply by maciejwalkowiak** [reply]
As far as I can tell, this is enough. @tomazfernandes if there's anything to add here maybe you can chime in

**Reply by andrej-urvantsev** [reply]
I did a small experiment: I stopped consumer, to accumulate some events in SQS(test env - low flow).

I tried to enable BATCH mode explicitly, just to check is there any difference:

```java
    @Bean
    MessageListenerContainer<String> listenerContainer(
            SqsAsyncClient sqsAsyncClient,
            @Value("${application.incoming-events-queue}") String queueNames,
            @Value("incoming_message_${INSTANCE_ID:${random.value}}") String id,
            IncomingEventsMessageListener incomingEventsMessageListener) {
        return SqsMessageListenerContainer.<String>builder()
                .configure(options -> options.listenerMode(ListenerMode.BATCH))
                .id(id)
                .sqsAsyncClient(sqsAsyncClient)
                .queueNames(queueNames)
                .messageListener(incomingEventsMessageListener)
                .build();
    }
```

Then I started it, there were 55 messages in SQS and logs looked like this(timestamp, thread name, batch size):

```
2024-09-19 14:00:26	incoming_message_subscription-webhook-sender-5db94c979d-h4948-1: incoming-events-queue batch size: 10
2024-09-19 14:00:27	incoming_message_subscription-webhook-sender-5db94c979d-h4948-2: incoming-events-queue batch size: 10
2024-09-19 14:00:27	incoming_message_subscription-webhook-sender-5db94c979d-h4948-3: incoming-events-queue batch size: 4
2024-09-19 14:00:27	incoming_message_subscription-webhook-sender-5db94c979d-h4948-4: incoming-events-queue batch size: 5
2024-09-19 14:00:27	incoming_message_subscription-webhook-sender-5db94c979d-h4948-5: incoming-events-queue batch size: 5
2024-09-19 14:00:27	incoming_message_subscription-webhook-sender-5db94c979d-h4948-6: incoming-events-queue batch size: 6
2024-09-19 14:00:27	incoming_message_subscription-webhook-sender-5db94c979d-h4948-7: incoming-events-queue batch size: 10
2024-09-19 14:00:27	incoming_message_subscription-webhook-sender-5db94c979d-h4948-8: incoming-events-queue batch size: 5
```

I would expect to see batch sizes like "10, 10, 10, 10, 10, 5"

Disclaimer: I'm very new to SQS world, so maybe I'm missing something obvious...

**Reply by maciejwalkowiak** [reply]
As far as I can tell there is no guarantee you'll always get 10 messages as they may come from different SQS shards (?). You can repeat similar experiment with pure SQS client - as far as I remember when I did that few years ago the results were the same.

**Reply by andrej-urvantsev** [reply]
Ok, I did a couple of more experiments and it seems that behavior depends of how many events are visible at the moment in SQS - more events in the queue, more consistent batches of 10.

Then the last question remain: I understand that it should be possible to configure batching somehow, that it does multiple polls, combine results and provides it to the consumer.

Does that mean that if I do something like this:
```java
    @SqsListener(
            queueNames = {"${application.incoming-events-queue}"},
            id = "incoming_message_${INSTANCE_ID:${random.value}}",
            maxConcurrentMessages = "20")
    public void handle(List<Message<String>> messages) {
```

and if there are enough messages visible in SQS, then SQS listener container will do 2 polls and invoke `handle` method with 20 messages?

**Reply by tomazfernandes** [reply, marked as answer]
Hi @lazystone, you're correct, if you set maxConcurrentMessages for a value greater than 10 it'll combine the result of multiple polls.

And as @maciejwalkowiak pointed out, that means you'll get messages up to that amount, but SQS does not guarantee to always deliver that amount.

And as you found out, the more messages in SQS, the higher the chances you'll receive full batches.

You don't need to manually configure BATCH mode - just having the List parameter will configure that for you.

**Reply by andrej-urvantsev** [reply]
Thanks for clarification!

**Reply by szykoc** [reply]
@tomazfernandes @maciejwalkowiak 
Is there a performance difference between the following two scenarios?

`public void handle(List<Message<String>> messages)`
`public void handle(Message<String> message)`

With the parameter:
`maxConcurrentMessages: 20`

Is it correct to say that:
1. The first one will poll 20 times, each time processing 10 messages?
2. The second one will poll 20 times, each time processing 1 message?

**Reply by tomazfernandes** [reply]
@szykoc, polling is completely decoupled from message handling logic, so whether you choose to receive the messages in a list or each one individually does not make any difference in how polling will behave.

In both cases, with default values, the framework will start by making 2 polls for 10 messages each.

With the List you'll receive all messages from both polls in the list (up to 20).

In the single message, you'll receive all messages from both polls in parallel (for standard queues).

Depending on your use case and business logic, processing itself might be more performant in one case or the other.

Does that answer your question?

**Reply by szykoc** [reply]
@tomazfernandes 

Thank you very much, it's clear right now :) 
Since we’re on the topic of performance, could you help me with tuning recommendations for my use case?

1. I'm processing around 50,000 messages per minute during business hours, with occasional peaks up to 70,000 messages. My SQSListener is currently running with default settings and acknowledgment mode set to ON_SUCCESS.
CPU usage stays around 15%, which is quite low. I assume I could increase the number of worker threads to boost throughput. Would that be the right approach?
The reason I’m asking is that during peak times, I notice that both `ApproximateAgeOfOldestMessage` and `ApproximateNumberOfMessagesVisible` metrics increase in SQS. So I assume that due to insufficient parallel processing it's is forming a backlog in my queue. 
Given that there are available CPU/MEM resources, would increasing the concurrency ex. worker threads be an effective way to reduce message latency and prevent queue buildup? With what values shall I start?  Also, are there other tuning options I should consider to improve performance under high load?

2. Separate question: I’m currently using ON_SUCCESS acknowledgment mode. However, in one of my use cases, when a specific business error occurs, I want to move the message to the DLQ and ensure it is removed from the main queue, so it’s not picked up again by another instance.
What’s the proper way to handle this scenario?

**Reply by szykoc** [reply]
@tomazfernandes could you please advise :)

**Reply by tomazfernandes** [reply]
Hey @szykoc, sorry, I missed getting back to you on this.

I'll try to give you a more complete answer by the end of the week, but the first setting you want to increase is the maxInflightMessages.

If you increase that setting, the framework will combine the result of many polls made in parallel, and process each message in a separate thread. You can try values as large as you'd like - 100, 1000, depending on your instance size, and check which value brings the best performance.

You can also look into other settings such as maxDelayBetweenPolls if you want extra-tuning, but that adds more complexity.

Depending on your business logic, you can also consider processing messages in batch.

For your second question regarding moving the message straight to DLQ based on exception classification, that's not supported by the framework at this point, but sounds interesting - would you be able to open a feature request for that? We can discuss it in more details there.

Please let me know if it makes sense to you.

**Reply by szykoc** [reply]
Hey @tomazfernandes 
Thank you, I appreciate your help :) I'm looking forward to more details on complex tuning!
In my case, I need to process messages as quickly as possible, since I consume and push events to business users, so any delays are not ideal.
However, my flow involves a few external service calls that I don't control. Currently, I use an `AsyncErrorHandler` which acknowledges everything even on error, by returning `CompletableFuture.completedFuture(null);` (I don’t need to reprocess messages).

**Reply by szykoc** [reply]
Hey @tomazfernandes hope all is good :) 

Have you had a chance to look into more complex solutions for increasing throughput? I’ve just started experimenting with `maxConcurrentMessages`.
However, I’m noticing some strange behavior in the SQS monitoring. Sometimes, messages aren't visible, yet the age of the oldest message in the queue (that hasn't been picked up by the listener) keeps increasing to around 9–10 seconds. I’m seeing this pattern repeatedly.
The threads don’t seem to be stuck in the business logic. Could this be related to `maxDelayBetweenPolls`?

![image](https://github.com/user-attachments/assets/24711f0e-bfde-40f6-9ca4-c9be310d2d8a)

**Reply by tomazfernandes** [reply]
Hey @szykoc.

As the image suggests, the value AWS provides is an approximation, so you might want to have a more reliable metric e.g. tracing / logs between the messages are produced vs. the time they're consumed.

One reason you might be seeing this is due to the `pollTimeout` setting. By default, the framework will wait up to 10 seconds for a full batch of messages to be retrieved. You can set that to e.g. 1 second, and then it'll return any messages present in the queue (or, more precisely, in the SQS instance it's polling from) in that time.

`maxDelayBetweenPolls` sets the time the framework will wait for a full batch of permits to be released before making another poll. This is useful for attempting to make polls for full batches versus making more polls for smaller number of messages.

You can also set `BackPressureMode` to `FIXED_HIGH_THROUGHPUT` - this will prevent the framework from entering `LOW_THROUGHPUT` mode when polls return empty. Low throughput mode is to prevent the framework from making numerous polls to AWS when the queue is not active.

If you want more insight on how the framework is behaving, you can tune up logging for classes such as `SemaphoreBackPressureHandler`, `AbstractPollingMessageSource`, `AbstractMessageProcessingPipelineSink`.

Note that these are all mostly tradeoffs of performance versus cost. If you tune all settings up for performance, it might increase the number of polls the framework makes to AWS, and it might incur in a higher bill.

Let me know if that works for you, thanks.
